SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Utils/UserIdContext.cs

```diff
@@ -48,7 +57,27 @@ public static UserIdContext Instance
         /// <summary>
         /// Current user identifier.
         /// </summary>
-        public string UserId { get; set; }
+        public string UserId
+        {
+            get
+            {
+                lock (UserIdContextLock)
+                {
+                    return _userId;
+                }
+            }
+            set
+            {
+                lock (UserIdContextLock)
+                {
+                    if (_userId == null || !_userId.Equals(value))
+                    {
+                        _userId = value;
+                        UserIdUpdated?.Invoke(this, new UserIdUpdatedEventArgs { userId = value });
```

Comment:
Invoking a callback inside of a lock is prone to deadlocking, consider saving a local variable of the callback from within the lock but invoking it outside.